test/test_demos.py

```diff
@@ -0,0 +1,248 @@
+from selenium.webdriver.support import expected_conditions as EC
+import multiprocessing
+import time
+import requests
+from matplotlib.testing.compare import compare_images
+import random
+import os
+
+current_dir = os.getcwd()
+
+LOCAL_HOST = "http://localhost:{}"
+GOLDEN_PATH = "test/golden/{}/{}.png"
+TOLERANCE = 0.1
+
+
+def wait_for_url(url):
+    for i in range(10):
+        try:
+            requests.get(url)
+            print("Interface connected.")
+            break
+        except:
+            time.sleep(0.2)
+    else:
+        raise ConnectionError("Could not connect to interface.")
+
+
+def hide_latency(driver):
+    js = "document.getElementsByClassName('loading_time')[" \
+         "0].style.visibility = " \
+         "'hidden';"
+    driver.execute_script(js)
+
+
+def diff_texts_thread(return_dict):
+    from demo.diff_texts import io
+    io.save_to = return_dict
+    io.launch()
+
+
+def image_mod_thread(return_dict):
+    from demo.image_mod import io
+    io.examples = None
+    io.save_to = return_dict
+    io.launch()
+
+
+def longest_word_thread(return_dict):
+    from demo.longest_word import io
+    io.save_to = return_dict
+    io.launch()
+
+
+def sentence_builder_thread(return_dict):
+    from demo.sentence_builder import io
+    io.save_to = return_dict
+    io.launch()
+
+
+class TestDemo(unittest.TestCase):
+    def test_diff_texts(self):
+        manager = multiprocessing.Manager()
+        return_dict = manager.dict()
+        self.i_thread = multiprocessing.Process(target=diff_texts_thread,
+                                         args=(return_dict,))
+        self.i_thread.start()
+        while not return_dict:
+            time.sleep(0.2)
+
+        URL = LOCAL_HOST.format(return_dict["port"])
+        wait_for_url(URL)
+
+        driver = webdriver.Chrome()
+        driver.set_window_size(1200, 800)
+        driver.get(URL)
+        timeout = 10
+        elem = WebDriverWait(driver, timeout).until(
+            EC.presence_of_element_located((By.CSS_SELECTOR,
+                                            ".input_interface[interface_id='0'] .input_text"))
+        )
+        elem.clear()
+        elem.send_keys("Want to see a magic trick?")
+        elem = WebDriverWait(driver, timeout).until(
+            EC.presence_of_element_located((By.CSS_SELECTOR,
+                                            ".input_interface[interface_id='1'] .input_text"))
+        )
+        elem.clear()
+        elem.send_keys("Let's go see a magic trick!")
+        elem = WebDriverWait(driver, timeout).until(
+            EC.presence_of_element_located((By.CSS_SELECTOR,
+                                            ".submit"))
+        )
+        elem.click()
+        elem = WebDriverWait(driver, timeout).until(
+            EC.presence_of_element_located((By.CSS_SELECTOR,
+                                            ".output_interface[interface_id='2'] .output_text"))
+        )
+        while elem.text == "":
```

Comment:
can we set a maximum time to this infinite while loops?

Reply:
Good point, added a time limit on this and all while loops in https://github.com/gradio-app/gradio/pull/72/commits/85c368387660a82c001ddb954b974e5d14826071